src/boolean3.cpp

```diff
@@ -448,44 +337,149 @@ struct Kernel12 {
+      tmp.v12_.push_back(v12, true);
+    }
+  }
+
+  Kernel12Tmp get() {
+#if MANIFOLD_PAR == 1
+    Kernel12Tmp result;
+    std::vector<Kernel12Tmp> tmps;
+    std::vector<SparseIndices> indices;
+    store.combine_each([&](Kernel12Tmp &data) {
+      indices.emplace_back(std::move(data.p1q2_));
+      tmps.emplace_back(std::move(data));
+    });
+    result.p1q2_.Clear();
+    result.p1q2_.FromIndices(indices);
+
+    std::vector<size_t> sizes;
+    size_t total_size = 0;
+    for (const auto &tmp : tmps) {
+      sizes.push_back(total_size);
+      total_size += tmp.x12_.size();
+    }
+    result.x12_.resize(total_size);
+    result.v12_.resize(total_size);
+    for_each_n(ExecutionPolicy::Seq, countAt(0), indices.size(), [&](size_t i) {
+      std::copy(tmps[i].x12_.begin(), tmps[i].x12_.end(),
+                result.x12_.begin() + sizes[i]);
+      std::copy(tmps[i].v12_.begin(), tmps[i].v12_.end(),
+                result.v12_.begin() + sizes[i]);
+    });
+    return result;
+#else
+    return localStore;
+#endif
+  }
+};
+
+std::tuple<Vec<int>, Vec<vec3>> Intersect12(const Manifold::Impl &inP,
+                                            const Manifold::Impl &inQ,
+                                            SparseIndices &p1q2, double expandP,
+                                            bool forward) {
   ZoneScoped;
-  Vec<int> x12(p1q2.size());
-  Vec<vec3> v12(p1q2.size());
+  F02 f02{inP.vertPos_,
+          inQ.halfedge_,
+          inQ.vertPos_,
+          expandP,
+          forward ? inP.vertNormal_ : inQ.vertNormal_,
+          forward};
+  F11 f11{forward ? inP.vertPos_ : inQ.vertPos_,
+          forward ? inQ.vertPos_ : inP.vertPos_,
+          forward ? inP.halfedge_ : inQ.halfedge_,
+          forward ? inQ.halfedge_ : inP.halfedge_,
+          expandP,
+          forward ? inP.vertNormal_ : inQ.vertNormal_};
+
+  // inQ.EdgeCollisions(inP)
+  Vec<TmpEdge> tmpedges = CreateTmpEdges(inP.halfedge_);
+  Vec<Box> PEdgeBB(tmpedges.size());
+  {
+    const size_t numEdge = tmpedges.size();
+    const auto &vertPos = inP.vertPos_;
+    auto policy = autoPolicy(numEdge, 1e5);
+    for_each_n(policy, countAt(0), numEdge,
+               [&PEdgeBB, &tmpedges, &vertPos](const int e) {
+                 PEdgeBB[e] = Box(vertPos[tmpedges[e].first],
+                                  vertPos[tmpedges[e].second]);
+               });
+  }
+  F12 f12{inP.halfedge_, inQ.halfedge_, inP.vertPos_, forward, f02, f11};
+  Kernel12Recorder recorder{f12, tmpedges, forward};
 
-  for_each_n(
-      autoPolicy(p1q2.size(), 1e4), countAt(0_uz), p1q2.size(),
-      Kernel12({x12, v12, p0q2.AsVec64(), s02, z02, p1q1.AsVec64(), s11, xyzz11,
-                inP.halfedge_, inQ.halfedge_, inP.vertPos_, forward, p1q2}));
+  inQ.collider_.Collisions<false, Box, Kernel12Recorder>(PEdgeBB.cview(),
+                                                         recorder);
 
-  p1q2.KeepFinite(v12, x12);
+  Kernel12Tmp result = recorder.get();
```

Comment:
This is the place where things is broken. p1q2's order is now non-deterministic, and it seems that `Sponge4` will randomly report

```
manifold/src/polygon.cpp (124): 'std::all_of(triangles.begin(), triangles.end(), [&vertPos, epsilon](const ivec3 &tri) { return CCW(vertPos[tri[0]], vertPos[tri[1]], vertPos[tri[2]], epsilon) >= 0; })' is false: triangulation is not entirely CCW!" thrown in the test body.
```

but I don't see anything wrong in the code apart from the different order. The major issue here is that the error is pretty rare (20%?), and never happens when I make the collider sequential, so it is probably related to our collider code. I am not sure if this is an old bug or a new one.

Reply:
I would guess it's triggering a bug in our triangulator, rather then being actually wrong. If you can manage to dump the polygon into a new test when it happens to fail, that would be ideal. However, we have lots of people asking us for determinism, so I think regardless we'll need to sort or something.

Reply:
ok, just precision issue